docs/experiments/join-mode/recursive-joins.md

```diff
@@ -0,0 +1,222 @@
+# Recursive joins
+
+Join mode is an extension on top of Semgrep that runs multiple rules at once and only returns results if certain conditions are met. Join mode is an experimental mode that lets you cross file boundaries, allowing you to write rules for whole code bases instead of individual files. More information is available in the docs, here: https://semgrep.dev/docs/experiments/join-mode/
+
+Recursive join mode adds a new recursive operator, `-->`, which executes a recursive query on the given condition. This lets you write a Semgrep rule that effectively crawls the codebase on some condition, letting you build chains such as function call chains or class inheritance chains.
+
+## How recursve join mode works
+
+Behind the scenes, join rules turn captured metavariables into database table columns. This means a rule with $FUNCTIONNAME, $FUNCTIONCALLED, and $PARAMETER would be a table that looks like this:
+
+| $FUNCTIONNAME | $FUNCTIONCALLED | $PARAMETER   |
+|---------------|-----------------|--------------|
+| getName       | writeOutput     | user         |
+| getName       | lookupUser      | uid          |
+| lookupUser    | databaseQuery   | uid          |
+
+The join conditions then join various tables together and return a result if any rows match the criteria.
+
+Recursive join mode conditions use [recursive joins](https://www.sqlite.org/lang_with.html#recursive_common_table_expressions) to construct a table that recursively joins with itself. For example, you can use a Semgrep rule that gets all function calls and join it recursively to approximate a callgraph.
+
+Consider the following Python script and rule.
+
+```python
+def function_1():
+    print("hello")
+    function_2()
+
+def function_2():
+    function_4()
+
+def function_3():
+    function_5()
+
+def function_4():
+    function_5()
+
+def function_5():
+    print("goodbye")
+```
+
+```yaml
+rules:
+- id: python-callgraph
+  message: python callgraph
+  languages: [python]
+  severity: INFO
+  pattern: |
+    def $CALLER(...):
+      ...
+      $CALLEE(...)
+```
+
+A join condition like this, `python-callgraph.$CALLER --> python-callgraph.$CALLEE`, will produce a table like this. Notice how `function_1` appears with `function_4` and `function_5` as callees, even though it isn't directly called.
+
+| $CALLER  | $CALLEE  |
+|----------|----------|
+|function_1|function_2|
+|function_1|function_4|
+|function_1|function_5|
+|function_1|print     |
+|function_2|function_4|
+|function_2|function_5|
+|function_3|function_5|
+|function_4|function_5|
+|function_5|print     |
+
+## Example Rule
+
+To really utilize join mode, it's important to think of a join mode rule as "asking questions about the whole project", rather than looking for a single pattern. For example, to find a SQL injection, you need to understand a few things about the project:
```

Comment:
```suggestion
It's important to think of a join mode rule as "asking questions about the whole project", rather than looking for a single pattern. For example, to find a SQL injection, you need to understand a few things about the project:
```